### DISCUSSION TITLE
allow to inherit imports from amended file

### DISCUSSION THREAD

**Discussion by lucsoft** [apple/pkl#865] [Ideas]
Allowing to have implicit imports would remove the boilerplate of splitting up files. 

Current use case would be to not need to reimport all the Kubernetes imports for each file.

**Comment by bioball** [comment]
I definitely see the reason you might want this. However, this can introduce quite hidden bugs. If the parent's imports were resolvable in an amending module, this means that you can possibly change the meaning of a variable in a child module by simply adding an import.

For example:

```pkl
// parent.pkl
class Bird {
  name: String
  scientificName: String
}

bird: Bird
```

```pkl
// child.pkl
amends "parent.pkl"

bird {
  scientificName = name // `name` is not in lexical scope, fall back to looking up `this.name`
}
```

Here, `name` resolves to an implicit `this.name` because `name` is not available in `child.pkl`'s lexical scope.

If we allowed the child module to see imports declared in the parent, that means that we can change what `scientificName = name` resolves to, by adding this import:

```diff
// parent.pkl

+import "name.pkl"

 class Bird {
   name: String
   scientificName: String
 }

 bird: Bird
```

This type of behavior can cause hidden bugs; adding a new import can unintentionally cause a variable's resolution to change.

**Reply by bioball** [reply]
Note: often, you don't need to re-import in amending modules. Let's use the same example from above, but move `Bird` into its own module.

```pkl
// parent.pkl
import "Bird.pkl"

bird: Bird

birds: Listing<Bird>
```

An amending module does not need to then re-import `Bird` to define an instance of it. Instead, you can simply amend values of type `Bird` directly, or create instances with the `new` keyword in positions where Pkl can infer the parent (like in the case of `Listing<Bird>`.

```pkl
amends "parent.pkl"

bird {
  name = "Warbler"
  scentificName = "Parulidae"
}

birds {
  new { name = "Warbler"; scientificName = "Parulidae" }
}
```

**Reply by lucsoft** [reply]
Yeah the problem starts when "Bird" is like an abstract module (as seen in K8sResource)

```pkl
amends "common.pkl"

resources {
   new Deployment {
    
   }
   new Ingress {
   
   }
   new Service {
   
   }
}
```

this would be really nice if that just worked (imports are coming from common.pkl and i don't need to re-import)

> Here, name resolves to an implicit this.name because name is not available in child.pkl's lexical scope.

I get that but that scope "issue" also happens in classes where this is not the class but something more nested (i feel like btw this behavior is weird anyway) but thats ouf of topic
